### PR TITLE
Bugfix/batch eq for scalar

### LINKED ISSUE #1182: Make Batch.__eq__ work also for scalar values (0-dimensional arrays)
As first noticed here https://github.com/thu-ml/tianshou/issues/1086#issuecomment-2157976997, DeepDiff which is currently used for batch equality cannot handle 0-dimensional arrays.

My current proposal is to leverage `np.atleast_1d` and the newest Batch API's `Batch.apply_values_transform` to convert any 0-dimensional arrays to 1-d while keeping the others as they are (cf. https://github.com/thu-ml/tianshou/issues/1086#issuecomment-2263079148).

### PR DESCRIPTION
**Pull request by dantp-ai**
Fixes: https://github.com/thu-ml/tianshou/issues/1182

Note: Updated `test_batch.test_slice_distribution()` to use allclose (See: https://github.com/thu-ml/tianshou/pull/1181).

### CODE REVIEW

File: test/base/test_batch.py [hunk 1]
```diff
@@ -837,10 +859,11 @@ def test_slice_distribution() -> None:
         selected_idx = [1, 3]
         sliced_batch = batch[selected_idx]
         sliced_probs = cat_probs[selected_idx]
-        assert (sliced_batch.dist.probs == Categorical(probs=sliced_probs).probs).all()
-        assert (
-            Categorical(probs=sliced_probs).probs == get_sliced_dist(dist, selected_idx).probs
-        ).all()
+        assert torch.allclose(sliced_batch.dist.probs, Categorical(probs=sliced_probs).probs)
```
Comment (MischaPanch) [inline test/base/test_batch.py:862]: Thx! I guess it was failing randomly, right?

**Review by MischaPanch — APPROVED** [review]

**Comment by MischaPanch** [comment]
Thx for the PR! All looks good, merging now

File: test/base/test_batch.py (hunk 1 above)
Comment (dantp-ai) [reply test/base/test_batch.py:862]: Yes, it was sporadically returning inequality. I will take a look at __getitem__ soon as I have an overdue task there with maintaining [slicing consistency](https://github.com/thu-ml/tianshou/issues/1090). Then I can also check what causes slicing to make these little decimal errors.

Comment (MischaPanch) [reply test/base/test_batch.py:862]: Thanks, appreciate it! I don't think the getitem issue can actually be solved, but implementing eq already solves all practical problems of it. So no need to further address it, unless you see some simple way for doing that. 

The order of slicing however it's still a valid issue

**Pull request merged**